### PR TITLE
TRUNK-5968: Add Liquibase snapshot files for openmrs-core 2.4.x

### PR DESCRIPTION
**Pull request by BartlomiejRasztabiga**
## Description of what I changed
Add Liquibase snapshot files for openmrs-core 2.4.x

## Issue I worked on
see https://issues.openmrs.org/browse/TRUNK-5968

## Checklist: I completed these to help reviewers :)
- [X] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [X] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [X] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [X] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

**Comment by BartlomiejRasztabiga** [comment]
@WolfSchlegel I can see that there are indexes created with duplicated names and that's why tests are failing. Is it an expected behavior of `create_liquibase_snapshots.sh` script and should I rename them manually?

**Comment by dkayiwa** [comment]
@WolfSchlegel does this look good to you? 😊

**Comment by dkayiwa** [comment]
@BartlomiejRasztabiga did you test these changes by running a fresh setup of the openmrs web application?

**Comment by BartlomiejRasztabiga** [comment]
@dkayiwa Yes, I've installed new OpenMRS instance with success:
image

**Comment by dkayiwa** [comment]
@BartlomiejRasztabiga without complaining bout the missing `liquibase-update-to-latest-2.5.x.xml`?

**Comment by BartlomiejRasztabiga** [comment]
@dkayiwa It's not added to `liquibase-update-to-latest-from-1.9.x.xml` so I guess it shouldn't be complaining. I'm worried though that's not how it's supposed to be done. But when I add this file as in the instruction with no change sets, there are some unit tests failing.
`ChangeLogVersionsTest.shouldGetUpdateVersions` - that's the one failing
I'm including the assertion error on pastebin:
https://pastebin.com/TS1Wh8bh

**Comment by WolfSchlegel** [comment]
Hey there, I am on sick leave this week and can have a look when I am back to work.

**Comment by BartlomiejRasztabiga** [comment]
@dkayiwa I've pushed my latest changes to reproduce test error that I'm facing after adding `liquibase-update-to-latest-2.5.x.xml`. You can read my explanation in the comment above.

**Comment by dkayiwa** [comment]
@WolfSchlegel sorry to hear that you are sick. Wish you a quick recovery!

**Comment by dkayiwa** [comment]
@BartlomiejRasztabiga does `mvn clean install` succeed locally?

**Comment by BartlomiejRasztabiga** [comment]
@dkayiwa This one test is failing:
`ChangeLogDetectiveDatabaseIT.shouldGetUnrunLiquibaseUpdateFileNames`
https://pastebin.com/rDwL9zxn
In ChangeLogDetective.getUnrunLiquibaseUpdateFileNames method, we're adding unrun changeset filename only if there are any changesets unrun and there are not change sets in liquibase-update-to-latest-2.5.x.xml.

**Comment by dkayiwa** [comment]
@BartlomiejRasztabiga for now, you can add this line in the failing test:
`Whitebox.setInternalState(ChangeLogVersions.class, "UPDATE_VERSIONS", Arrays.asList("1.9.x", "2.0.x", "2.1.x", "2.2.x", "2.3.x", "2.4.x"));`

Remember to also include a TODO comment that this line should be removed once we have at least a changeset in the `liquibase-update-to-latest-2.5.x.xml` file.

**Comment by BartlomiejRasztabiga** [comment]
@dkayiwa I've noticed that I also have to add following annotations to make Whitebox work:
```
@RunWith(PowerMockRunner.class)
@PrepareForTest( { ChangeLogVersions.class })
```
The thing is that I cannot import `PowerMockRunner` from `org.powermock.modules.junit4.PowerMockRunner`. I get compilation error: `java: package org.powermock.modules.junit4 does not exist`.
It's weird as I can see the `org.powermock.modules.junit4.PowerMockRunner `class on the classpath

**Comment by dkayiwa** [comment]
@BartlomiejRasztabiga i was getting confused by your error because it works perfectly fine for me. 😊

File: api/src/main/resources/org/openmrs/liquibase/updates/liquibase-update-to-latest-2.5.x.xml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+<databaseChangeLog
+	xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext"
+	xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog-ext 
+	    http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd
+		http://www.liquibase.org/xml/ns/dbchangelog 
+		http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-2.0.xsd">
+	
+	<!-- This changelog is empty on purpose. It is used by the org.openmrs.util.DatabaseUpdater class. -->
```
Comment (dkayiwa) [inline api/src/main/resources/org/openmrs/liquibase/updates/liquibase-update-to-latest-2.5.x.xml:22]: I find the above comment confusing.

Comment (BartlomiejRasztabiga) [reply api/src/main/resources/org/openmrs/liquibase/updates/liquibase-update-to-latest-2.5.x.xml:22]: It's a copy of `liquibase-empty-changelog.xml` as said in the instruction.

Comment (BartlomiejRasztabiga) [reply api/src/main/resources/org/openmrs/liquibase/updates/liquibase-update-to-latest-2.5.x.xml]: Deleted the comment in https://github.com/openmrs/openmrs-core/pull/3728/commits/763952777aa8c08e10fc037413aeb2980d7fc8e2

**Pull request merged**